src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/ExtractToNewComponentCodeActionProvider.cs

```diff
@@ -0,0 +1,113 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Components;
+using Microsoft.AspNetCore.Razor.Language.Extensions;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+using Microsoft.AspNetCore.Razor.LanguageServer.CodeActions.Models;
+using Microsoft.AspNetCore.Razor.Threading;
+using Microsoft.CodeAnalysis.Razor.Logging;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.CodeActions.Razor;
+
+internal sealed class ExtractToNewComponentCodeActionProvider : IRazorCodeActionProvider
+{
+    private readonly ILogger _logger;
+
+    public ExtractToNewComponentCodeActionProvider(ILoggerFactory loggerFactory)
+    {
+        _logger = loggerFactory.GetOrCreateLogger<ExtractToNewComponentCodeActionProvider>();
+    }
+
+    public Task<IReadOnlyList<RazorVSInternalCodeAction>?> ProvideAsync(RazorCodeActionContext context, CancellationToken cancellationToken)
+    {
+        if (context is null)
+        {
+            return SpecializedTasks.Null<IReadOnlyList<RazorVSInternalCodeAction>>();
+        }
+
+        if (!context.SupportsFileCreation)
+        {
+            return SpecializedTasks.Null<IReadOnlyList<RazorVSInternalCodeAction>>();
+        }
+
+        if (!FileKinds.IsComponent(context.CodeDocument.GetFileKind()))
+        {
+            return SpecializedTasks.Null<IReadOnlyList<RazorVSInternalCodeAction>>();
+        }
+
+        var syntaxTree = context.CodeDocument.GetSyntaxTree();
+        if (syntaxTree?.Root is null)
+        {
+            return SpecializedTasks.Null<IReadOnlyList<RazorVSInternalCodeAction>>();
+        }
+
+        var owner = syntaxTree.Root.FindInnermostNode(context.Location.AbsoluteIndex, true);
```

Comment:
Please add names for boolean literal arguments for readability. In this case, I _guess_ it would be `includeWhitespace: true`, but I'm not 100% sure. Without opening your branch in Visual Studio and hovering over `FindInnermostNode(...)`, it's impossible to know.